lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -2781,12 +2787,397 @@ ModulePass *llvm::createDxilAllocateResourcesForLibPass() {
+      }
+    }
+  }
+  // propagate CreateHandleFromHeapArgs for HandleSel which all operands are
+  // other HandleSel.
+  void PropagateHeapArgs() {
+    SmallVector<Value *, 4> Candidates;
+    for (auto &Select : HandleSelects) {
+      CreateHandleFromHeapArgs &args = HandleToArgs[Select];
+      if (args.isResolved)
+        continue;
+      Candidates.emplace_back(Select);
+    }
+
+    while (1) {
+      SmallVector<Value *, 4> NextPass;
+      for (auto &Select : Candidates) {
+        CreateHandleFromHeapArgs &args = HandleToArgs[Select];
+        if (PHINode *Phi = dyn_cast<PHINode>(Select)) {
+          mergeHeapArgs(Phi, args.Index, Phi->incoming_values());
+        } else if (SelectInst *Sel = dyn_cast<SelectInst>(Select)) {
+          User::op_range range = User::op_range(Sel->getOperandList() + 1,
+                                                Sel->getOperandList() + 3);
+          mergeHeapArgs(
+              Sel, args.Index,
+              range);
+        } else {
+          DXASSERT(false, "otherwise, non-select/phi in Selects set");
+        }
+        // Reload args incase it is changed in mergeHeapArgs.
+        args = HandleToArgs[Select];
+        if (args.isResolved)
+          continue;
+        NextPass.emplace_back(Select);
+      }
+      // Some node cannot be reached.
+      if (NextPass.size() == Candidates.size())
+        return;
+      Candidates = NextPass;
+    }
+  }
+
+  void UpdateSelectsForHandleSelect(hlsl::OP *hlslOP) {
+    if (HandleSelects.empty())
+      return;
+    LLVMContext &Ctx = HandleSelects[0]->getContext();
+    Type *pVoidTy = Type::getVoidTy(Ctx);
+    // NOTE: currently phi of createHandleFromHeap and createHandleFromBinding
+    // is not supported.
+    Function *createHdlFromHeap =
+        hlslOP->GetOpFunc(DXIL::OpCode::CreateHandleFromHeap, pVoidTy);
+    Value *hdlFromHeapOP = hlslOP->GetI32Const(
+        static_cast<unsigned>(DXIL::OpCode::CreateHandleFromHeap));
+    for (auto &Select : HandleSelects) {
+      if (PHINode *Phi = dyn_cast<PHINode>(Select)) {
+        unsigned numIncoming = Phi->getNumIncomingValues();
+        CreateHandleFromHeapArgs &args = HandleToArgs[Phi];
+        PHINode *newPhi = cast<PHINode>(args.Index);
+        if (args.isResolved) {
+          for (unsigned j = 0; j < numIncoming; j++) {
+            // Set incoming values to undef until next pass
+            Value *V = Phi->getIncomingValue(j);
+            auto it = HandleToArgs.find(V);
+            DXASSERT(it != HandleToArgs.end(),
+                     "args.isResolved should be false");
+            CreateHandleFromHeapArgs &itArgs = it->second;
+            newPhi->setIncomingValue(j, itArgs.Index);
+          }
+
+          IRBuilder<> B(Phi->getParent()->getFirstNonPHI());
+          B.SetCurrentDebugLocation(Phi->getDebugLoc());
+          Value *isSampler = hlslOP->GetI1Const(args.isSampler);
+          // TODO: or args.IsNonUniform with !isUniform(Phi).
```

Comment:
Or emit warning or error if `(isUniform(Phi) && !args.IsNonUniform)`, since we should encourage explicit nonuniform marking and sinking of indexing in these cases, rather than silently try to fix it for them with performance implications.